nav2_simple_commander/nav2_simple_commander/robot_navigator.py

```diff
@@ -619,34 +666,48 @@ def smoothPath(
         self, path, smoother_id='', max_duration=2.0, check_for_collision=False
     ):
         """Send a `SmoothPath` action request."""
+        self.clearLastError()
         rtn = self._smoothPathImpl(path, smoother_id, max_duration, check_for_collision)
 
-        if self.status != GoalStatus.STATUS_SUCCEEDED:
-            self.warn(f'Getting path failed with status code: {self.status}')
-            return None
-
-        if not rtn:
-            return None
-        else:
+        if self.status == GoalStatus.STATUS_SUCCEEDED:
             return rtn.path
+        else:
+            self.setLastError(rtn.error_code, rtn.error_msg)
+            self.warn('Getting path failed with status code:'
+                      f'{self.status}:{rtn.error_code}:{rtn.error_msg}')
+            return None
 
-    def changeMap(self, map_filepath):
+    def changeMap(self, map_filepath) -> bool:
         """Change the current static map in the map server."""
+        self.clearLastError()
         while not self.change_maps_srv.wait_for_service(timeout_sec=1.0):
             self.info('change map service not available, waiting...')
         req = LoadMap.Request()
         req.map_url = map_filepath
         future = self.change_maps_srv.call_async(req)
         rclpy.spin_until_future_complete(self, future)
-        status = future.result().result
-        if status != LoadMap.Response().RESULT_SUCCESS:
-            self.error('Change map request failed!')
+        result = future.result().result
+        if result != LoadMap.Response().RESULT_SUCCESS:
+            if result == LoadMap.RESULT_MAP_DOES_NOT_EXIST:
+                reason = 'Map does not exist'
+            elif result == LoadMap.INVALID_MAP_DATA:
+                reason = 'Invalid map data'
+            elif result == LoadMap.INVALID_MAP_METADATA:
+                reason = 'Invalid map metadata'
+            elif result == LoadMap.UNDEFINED_FAILURE:
+                reason = 'Undefined failure'
+            else:
+                reason = 'Unknown'
+            self.setLastError(result, reason)
+            self.error(f'Change map request failed:{reason}!')
+            return False
         else:
             self.info('Change map request was successful!')
-        return
+            return True
 
     def clearAllCostmaps(self):
         """Clear all costmaps."""
+        self.clearLastError()
```

Comment:
This calls services, these can't populate error codes

Reply:
Changed api to  set|get|clear LastActionError()  and removed from services.